Rename onCheckoutBeforeProcessing to onCheckoutValidationBeforeProcessing

ralucaStan opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This commit also adds a deprecation warning when using onCheckoutBeforeProcessing.
See the issue below for details.

<!-- Reference any related issues or PRs here -->
Fixes #3984

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->


### How to test the changes in this Pull Request:
The renaming part could be tested by smoke testing the Checkout Block:

1. Add something to checkout
2. Submit without a required field
3. Observe validation happens
4. Fix validation and place order

To see the warning you have to call yourself the old `onCheckoutBeforeProcessing`. What I did:

1. I added this component to [Order Summary](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/base/components/cart-checkout/order-summary/index.js) and included it as a child.
2. I went to the Checkout block and tried to submit it without some required fields.
3. I checked the console for the deprecation warning
4. I then fixed the warnings and  successfully submitted the order



```js
const Component = () => {
	const { onCheckoutBeforeProcessing } = useCheckoutContext();
	useEffect( () => {
		const unsubscribe = onCheckoutBeforeProcessing( () => {
			console.log( 'My custom validation callback' );
			return true;
		} );
		return unsubscribe;
	}, [ onCheckoutBeforeProcessing ] );
	return null;
};
```

One thing that could also help with testing is checking the registered observers by adding `console.log(observersByType)` in [emitters.ts](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/base/context/shared/event-emit/emitters.ts#L26). This is logged after submitting with missing inputs.

<!-- If you can, add the appropriate labels -->

### Changelog

> Rename onCheckoutBeforeProcessing to onCheckoutValidationBeforeProcessing.

### Dev note

> `onCheckoutBeforeProcessing` has been renamed to `onCheckoutValidationBeforeProcessing` to more clearly articulate its coupling to the validation happening when the Checkout form is submitted. `onCheckoutBeforeProcessing` will be deprecated in the future.

ralucaStan committed:
Rename onCheckoutBeforeProcessing to onCheckoutValidationBeforeProcessing

This commit also adds a deprecation warning when using onCheckoutBeforeProcessing.

opr commented on assets/js/type-defs/contexts.js:
```diff
@@ -120,16 +120,16 @@
  * @property {function(function(),number=)} onCheckoutAfterProcessingWithError   Used to register a callback that will
  *                                                                               fire when the checkout has been
  *                                                                               processed and has an error.
- * @property {function(function(),number=)} onCheckoutBeforeProcessing           Used to register a callback that will
```
Removing this causes a TypeScript warning - because this is still used in `usePaymentMethodInterface` but according to the type-defs does not exist on `useCheckoutContext`.

ralucaStan replied:
I prefer not to handle this here, I've added a deprecated comment where usePaymentMethodInterface is declared, once that file is converted to .ts I image it should be fine

opr reviewed (approved):
Looks great! Works well too, I just have one (non-blocking) comment about TypeScript warnings - it's up to you to decide if we care about this warning for now, or if you want to add the property back to the type-def.

ralucaStan committed:
Add deprecated comment

ralucaStan committed:
Prevent deprecated message from being called on initial render

Because  the function passed to useMemo runs during rendering the deprecation message was being called on render time.
This was also causing the e2e tests to fail.
This commit introduces another function between useMemo and the emitterObservers so that the deprecated message gets shown only at invocation time.

ralucaStan commented:
I fixed the tests with commit https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4039/commits/a502d1be8828848270993264c64220731b0ae181, see the description for details.

PR ready to retest @opr

opr reviewed (changes requested):
The latest change looks good, clever way of solving it! I know you did add the comment in the commit message about the reasoning behind this, but could you add an inline comment explaining it too? Without knowing the context this code looks pretty strange.

ralucaStan committed:
Add comment to explain onCheckoutValidationBeforeProcessing implementation

ralucaStan commented:
> The latest change looks good, clever way of solving it! I know you did add the comment in the commit message about the reasoning behind this, but could you add an inline comment explaining it too? Without knowing the context this code looks pretty strange.

Good idea, I added a comment.

opr commented on assets/js/base/context/providers/cart-checkout/checkout-state/index.js:
```diff
@@ -122,6 +122,11 @@ export const CheckoutStateProvider = ( {
 	/**
 	 * @deprecated use onCheckoutValidationBeforeProcessing instead
+	 *
+	 * To prevent the deprecation message to be shown at render time
```
```suggestion
	 * To prevent the deprecation message being shown at render time
```

opr commented on assets/js/base/context/providers/cart-checkout/checkout-state/index.js:
```diff
@@ -122,6 +122,11 @@ export const CheckoutStateProvider = ( {
+	 * To prevent the deprecation message to be shown at render time
+	 * we need an extra function between useMemo and emitterObservers
+	 * so that the deprecated message gets shown only at invocation time.
+	 * (useMemo calls the passed function at render time)
```
```suggestion
	 * (useMemo calls the passed function at render time)
	 * See: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4039/commits/a502d1be8828848270993264c64220731b0ae181
```

opr reviewed (approved):
Approved again! There's one grammar change and I would recommend adding the commit link for additional context. 🥳

ralucaStan committed:
Update assets/js/base/context/providers/cart-checkout/checkout-state/index.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Update assets/js/base/context/providers/cart-checkout/checkout-state/index.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
